### PR TITLE
feat: find correct child components' definition inside template region

### PR DESCRIPTION
**Pull request by hikerpig**
NOTE: This is a copy of PR [1760](https://github.com/vuejs/vetur/pull/1760), sorry that I republish original branch fixing conflicts but got messed up by github's workflow, had to create a new PR here.

This PR will add support for #1712.

I notice that there is already vueFileInfo.componentInfo.childComponents provided by the implementation of Vue interpolation. But during definition-finding for html, a little case issue causes this functionality to fail, typically when component filenames adopt a PascalCase convention.

### CODE REVIEW

File: test/lsp/features/definition/basic.test.ts [hunk 1]
```diff
@@ -12,22 +12,27 @@ describe('Should find definition', () => {
...
   it('finds definition for Vue#data', async () => {
     const vueOptionsDtsUri = getDocUri('node_modules/vue/types/options.d.ts');
-    await testDefinition(docUri, position(20, 2), sameLineLocation(vueOptionsDtsUri, 73, 2, 6));
+    await testDefinition(docUri, position(24, 2), sameLineLocation(vueOptionsDtsUri, 73, 2, 6));
   });
 
   it('finds definition for imported Vue files', async () => {
-    const itemUri = getDocUri('definition/Basic.Item.vue');
-    await testDefinition(docUri, position(17, 7), location(itemUri, 5, 0, 7, 1));
+    const itemUri = getDocUri('client/definition/Basic.Item.vue');
```
Comment (octref) [inline test/lsp/features/definition/basic.test.ts:29]:
```suggestion
    const itemUri = getDocUri('definition/Basic.Item.vue');
```

File: test/lsp/features/definition/basic.test.ts [hunk 2]
```diff
@@ -12,22 +12,27 @@ describe('Should find definition', () => {
...
+    await testDefinition(docUri, position(18, 7), location(itemUri, 5, 0, 7, 1));
+  });
+
+  it('finds definition for child component inside template region', async () => {
+    const itemUri = getDocUri('client/definition/Basic.Item.vue');
```
Comment (octref) [inline test/lsp/features/definition/basic.test.ts:34]:
```suggestion
    const itemUri = getDocUri('definition/Basic.Item.vue');
```

**Comment by hikerpig** [comment]
Sorry for the LSP tests paths, long time after my first commit. I've merged the master branch and fixed some conflicts. 
If everything is good, I wish to rebase and squash these commits before submit 😃

**Comment by octref** [comment]
Seems covered by #2096, sorry.

**Pull request closed without merging**